pennylane/fourier/advanced_spectrum.py

```
@@ -0,0 +1,365 @@
+    "x": {(0,): [-0.5, 0.0, 0.5], (1,): [-0.5, 0.0, 0.5], (2,): [-0.5, 0.0, 0.5]}
+    "y": {(0,): [-2.3, 0.0, 2.3], (1,): [-2.3, 0.0, 2.3], (2,): [-2.3, 0.0, 2.3]}
+    "z": {(): [-3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0]}
+
+    .. note::
+        While the Fourier spectrum usually does not depend
+        on trainable circuit parameters or the actual values of the inputs,
+        it may still change based on inputs to the QNode that alter the architecture
+        of the circuit.
+
+    Above, we selected all input-encoding parameters for the spectrum computation, using
+    the ``argnum`` keyword argument. We may also restrict the full analysis to a single
+    QNode argument, again using ``argnum``:
+
+    >>> res = advanced_spectrum(circuit, argnum=[0])(x, y, z, w)
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "x": {(0,): [-0.5, 0.0, 0.5], (1,): [-0.5, 0.0, 0.5], (2,): [-0.5, 0.0, 0.5]}
+
+    Selecting arguments by name instead of index is possible via the
+    ``encoding_args`` argument:
+
+    >>> res = advanced_spectrum(circuit, encoding_args={"y"})(x, y, z, w)
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "y": {(0,): [-2.3, 0.0, 2.3], (1,): [-2.3, 0.0, 2.3], (2,): [-2.3, 0.0, 2.3]}
+
+    Note that for array-valued arguments the spectrum for each element of the array
+    is computed. A more fine-grained control is available by passing index tuples
+    for the respective argument name in ``encoding_args``:
+
+    >>> encoding_args = {"y": [(0,),(2,)]}
+    >>> res = advanced_spectrum(circuit, encoding_args=encoding_args)(x, y, z, w)
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "y": {(0,): [-2.3, 0.0, 2.3], (2,): [-2.3, 0.0, 2.3]}
+
+    .. warning::
+        The ``advanced_spectrum`` function does not check if the result of the
+        circuit is an expectation value. It checks whether the classical preprocessing between
```

Wait. _Should_ we check that the output is an expectation value?

---

I'd vote for no, because it might be hidden that the QNode actually computes an expval. For example, returning `qml.probs(wires=dev.wires)[0]` does is not a `qml.expval` but measures the projector onto $|0>$.

---

I suppose we could just raise an error if the QNode contains `var` or `sample`?

---

Fair enough :) Will do!